homeassistant/components/openuv/coordinator.py

```diff
@@ -1,23 +1,94 @@
 """Define an update coordinator for OpenUV."""
 from __future__ import annotations
 
+import asyncio
 from collections.abc import Awaitable, Callable
 from typing import Any, cast
 
-from pyopenuv.errors import OpenUvError
+from pyopenuv.errors import InvalidApiKeyError, OpenUvError
 
-from homeassistant.core import HomeAssistant
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
 from homeassistant.helpers.debounce import Debouncer
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import LOGGER
+from .const import DOMAIN, LOGGER
 
 DEFAULT_DEBOUNCER_COOLDOWN_SECONDS = 15 * 60
 
 
+class InvalidApiKeyMonitor:
+    """Define a monitor for failed API calls (due to bad keys) across coordinators."""
+
+    DEFAULT_FAILED_API_CALL_THRESHOLD = 5
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self._count = 1
+        self._lock = asyncio.Lock()
+        self._reauth_flow_manager = ReauthFlowManager(hass, entry)
+        self.entry = entry
+
+    async def async_increment(self) -> None:
+        """Increment the counter."""
+        LOGGER.debug("Invalid API key response detected (number %s)", self._count)
+        async with self._lock:
+            self._count += 1
+            if self._count > self.DEFAULT_FAILED_API_CALL_THRESHOLD:
+                self._reauth_flow_manager.start_reauth()
+
+    async def async_reset(self) -> None:
+        """Reset the counter."""
+        async with self._lock:
+            self._count = 0
+            self._reauth_flow_manager.cancel_reauth()
+
+
+class ReauthFlowManager:
+    """Define an OpenUV reauth flow manager."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self.entry = entry
+        self.hass = hass
+
+    @callback
+    def _get_active_reauth_flow(self) -> FlowResult | None:
```

Comment:
I think the typing on this should be a `config_entries.ConfigFlow`?

Reply:
`async_progress_by_handler` returns `list[FlowResult]`: https://github.com/home-assistant/core/blob/785cf0e29caa7aea1e5b9cbc0e02e1cd9515cac1/homeassistant/data_entry_flow.py#L194-L201

Reply:
Absolutely it does.  Sorry about that.  I didn't context switch well from looking at 81733.